File: python-learning-path/practise/bubble_sort.py
"""
Exercise 11: Bubble Sort

ALGORITHM:
Iteratively compare adjacent elements and swap them if they are in the wrong order. 
With each full pass, the largest remaining element "bubbles" to the correct position.
"""

def bubble_sort(arr):
    n = len(arr)
    
    # Outer loop: controls the number of passes
    for i in range(n):
        # Optimization flag: if no swaps occur, the list is already sorted
        swapped = False
        
        # Inner loop: compares adjacent items
        # range(n - i - 1) because the last 'i' elements are already in place
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                # Swap elements
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        
        # If no two elements were swapped by inner loop, then break
        if not swapped:
            break

def main():
    arr = [64, 34, 25, 12, 22, 11, 90]
    print(f"Unsorted List: {arr}")
    
    bubble_sort(arr)
    
    print(f"Sorted List:   {arr}")

if __name__ == "__main__":
    main()

File: java-learning-path/practise/BubbleSort.java
import java.util.Arrays;

/**
 * Exercise 11: Bubble Sort
 * 
 * ALGORITHM (Sinking Sort):
 * In each pass, the largest element "bubbles up" to its correct position at the end.
 * Time Complexity: O(n^2) - Nested loops.
 */
public class BubbleSort {
    public static void main(String[] args) {
        int[] arr = {64, 34, 25, 12, 22, 11, 90};
        System.out.println("Unsorted: " + Arrays.toString(arr));

        bubbleSort(arr);

        System.out.println("Sorted:   " + Arrays.toString(arr));
    }

    public static void bubbleSort(int[] arr) {
        int n = arr.length;
        // Outer loop controls number of passes
        for (int i = 0; i < n - 1; i++) {
            
            // Inner loop does the comparisons and swaps
            // Optimization: 'n - i - 1' because the last 'i' elements are already sorted
            for (int j = 0; j < n - i - 1; j++) {
                
                // Swap if the current element is greater than the next one
                if (arr[j] > arr[j + 1]) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }
    }
}
